adding spack diff command

vsoch opened the pull request:
This PR will add a new spack command, "diff" that will take two specs, and then use the spack.solver.asp.SpackSolverSetup to generate lists of facts about each (e.g., nodes, variants, etc.) and then take a set difference between the two to show the user the differences. The use case for this is whenever a user has a disambiguate situation and cannot remember how two different installs are different. The command can also output `--json` in the case of a more analysis type use case where we want to save complete data with all diffs and the intersection. However, the command is really more intended for a command line use case, and we likely will have an analyzer more suited to saving data. A quick example from the added docs here:

```bash
$ spack diff /efzjziy /sl7m27m
==> diff(zlib@1.2.11/efzjziy, zlib@1.2.11/sl7m27m)
VARIANT_SET
  zlib optimize bool(False)
==> diff(zlib@1.2.11/sl7m27m, zlib@1.2.11/efzjziy)
VARIANT_SET
  zlib optimize bool(True)
```

This "diff" command is one of two additions that will help users to understand how packages can change over time. The use cases are the following:

 - **This PR**: "I want to know right now how these specs are different in the command interface"
 - **Diff analyzer**: I want to calculate a bunch of differences to export for analysis or visualization"

The similarity between the two will be the data structure (generated here with `--json` - this data structure is what will also be one of the few generated for an analyzer that I'll add that is specifically to compare specs over time. It can be plugged into interfaces like the one that I'm prototyping now: 

Since the analyzer needs the analyzer work from the spack-monitor PR, I'm undecided if I'll wait for that, or add the analyzer in a separate PR from spack monitor (if spack monitor isn't going to be merged for a while). 

### Update 3/15
The example now shows just using the hash, and we print one attribute per line.

```bash
$ spack diff /efzjziy /sl7m27m
==> diff(zlib@1.2.11/efzjziy, zlib@1.2.11/sl7m27m)
VARIANT_SET
zlib optimize bool(False)
==> diff(zlib@1.2.11/sl7m27m, zlib@1.2.11/efzjziy)
VARIANT_SET
zlib optimize bool(True)
```

### Update 3/18

Added spacing so the titles are different from content below them:

```bash
$ spack diff /efzjziy /sl7m27m
==> diff(zlib@1.2.11/efzjziy, zlib@1.2.11/sl7m27m)
VARIANT_SET
  zlib optimize bool(False)
==> diff(zlib@1.2.11/sl7m27m, zlib@1.2.11/efzjziy)
VARIANT_SET
  zlib optimize bool(True)
```

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding spack diff command

A spack diff will take two specs, and then use the spack.solver.asp.SpackSolverSetup to generate
lists of facts about each (e.g., nodes, variants, etc.) and then take a set difference between the
two to show the user the differences. The use case for this is whenever a user has a disambiguate
situation and cannot remember how two different installs are different. The command can also output
--json in the case of a more analysis type use case where we want to save complete data with all
diffs and the intersection. However, the command is really more intended for a command line use
case, and we likely will have an analyzer more suited to saving data

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
pytest imported but unused!

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
Looks like we need more tests for the diff command! I added just a few simple ones to test the comparison function, but neither of them uses the command line (e.g., calling SpackCommand `diff(*args)`, the reason being the disambiguate function requires that the specs be installed, and when I tried the same mock approach as I saw in load with mpileaks it didn't register them as installed. Tips/feedback for how to do this correctly would be appreciated!

vsoch commented:
oh I figured it out! Nevermind! WIll add more tests soon!

vsoch committed:
adding tests for diff that use command directly to test functions for json, first, bold

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

haampie commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,170 @@
+        dest='dump_json',
+        help="Dump json output instead of pretty printing."
+    )
+    subparser.add_argument(
```
I'm not sure about this flag, since there are other commands that currently just error when multiple specs match. Isn't the first result entirely arbitrary?

I'm thinking it would be nice to have a simple CLI menu option when spack is run in an interactive shell so the user can select a specific match. That could be used in `spack uninstall` when `--all` is not provided; as well as selecting one of the > 1 top-level specs from an environment when running `spack -e my_env build-env -- bash`, etc.

vsoch replied:
The parameter `--first` is used to disambiguate the spec - so for example, if I have multiple versions of mpileaks and I ask to diff that against something else, it will raise that disambiguation error and tell me to choose one. If I select first, then it just grabs the first result and gives me the difference, which (as a lazy person sometimes) is what I want. So I don't think it's entirely useless. I verified what I just walked through in testing.

Your second comment is an interesting idea, but it's out of scope for this PR. I'm not doing anything with an interactive spack shell, nor am I touching uninstall or similar. If it's something that could be useful for diff, maybe you could write it up in an issue and i can take a look after this PR?

vsoch replied:
We also need the disambiguate command because it's going to show the user the different options they have in detail (hashes, etc.) which is exactly what they need if they want to diff similar specs.

Update: I added a few more lines to the test file to cover the case of != 2 specs provided.

haampie replied:
I don't see how `--first` is useful though, since as a user you are probably uninformed about the order in which spack sorts matching specs, and it's not helpful if your spec is not listed first. I would probaby run the command without `--first` by default, and when it fails and outputs a bunch of matches, I would run it again but using the hash from the previous output; that's about as much work as writing `--first` and more flexible?

And about a menu option: it's definitely out of the scope of this PR, but I just commented to point out this multiple-spec-match-error is a recurring issue across different commands and we can make the user experience better.

vsoch replied:
If that argument is valid, then we would need to remove it also from load.py, where the user is also in the same boat. I added it here based on the same logic that it is there.  I just looked at the commit history, and I think the automated use case is a valid one: https://github.com/spack/spack/pull/15622. spack diff could be run in an automated fashion just the same as spack load.

haampie replied:
Hm yeah, but that use case is about loading an arbitrary version of a build dependency where any would suffice. In this case you want to compare two very specific things, not an arbitrary first match. (FWIW, I don't feel very strongly about this, maybe that's not clear over the internet...)

vsoch replied:
I also don't feel strongly enough to argue to keep it if the majority think its not useful. Could we ask for others' opinions and get more feedback? If more folks are in agreement with you I'm definitely okay just removing it. I do think there is a use case, but you might be right that we shouldn't develop for this use case until it actually presents itself.

tgamblin replied:
The use case I see for `--first` here is testing -- it just guarantees the thing succeeds *if it can*, which is sometimes what you want.  I don't think someone would use it in anger -- if I had to guess, `spack diff` will nearly always be used interactively and people will want to diff specific things.  I don't think it hurts to keep it, either, at least for now.

The case I was kind of thinking about where `--first` may cease to make a lot of sense is if you want to do more general queries.  I could see someone saying:

```console
$ spack diff python
```

and getting diffs among *all* the python installations.  That's a multi-way diff, something this doesn't currently do (and probably shouldn't in the first PR).

So my take is keep it for now because who cares and it helps @vsoch, and we'll see if it stands up as we add more to `spack diff`.

vsoch committed:
ensure we test case of having just one or >2 specs (not allowed)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
I keep seeing this error and I'm not sure if it's transient:

```

lib/spack/spack/test/llnl/util/tty/log.py:303: AssertionError
----------------------------- Captured stderr call -----------------------------
minion: stdin.isatty(): True
minion: ready!
pid:        25041
pgid:       25041
sid:        25041
tcgetpgrp:  25041

minion pid:  25085
minion pgid: 25085
minion sid:  25041

------------------------------------------- bg
canon: on, echo: on
input: False
bg: True

------------------------------------------- fg
canon: on, echo: on
input: False
bg: False
```
https://github.com/spack/spack/pull/22283/checks?check_run_id=2107486935

vsoch commented:
These CI runs errors are kind of random and spurious - it always seems to be one of the clingo versions that has that random error. I guess I should just keep re-running? it seems like not a great use of resources (at least I feel terrible about it) :disappointed:

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+
+
+Let's now say that we want to uninstall zlib. We run the command, and hit a problem
+real quickly that we have two!
```
```suggestion
real quickly since we have two!
```

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+
+Oh no! This is a good use case for ``spack diff``, which can easily show us the
+"diff" or set difference between properties for two packages. Let's try it out.
+Since the only difference between these two is the hash, we provide the hash.
```
Might want to clarify that the only distinction between the two entries shown here is the hash.

vsoch replied:
Gotcha.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+
+.. code-block::console
+
+    $ spack diff zlib@1.2.11/efzjziy zlib@1.2.11/sl7m27m
```
This may be premature as I work my way through the diffs, but could it be enough to just write the hashes, i.e.,:

```
$ spack diff /efzjziy /sl7m27m
```

or the equivalent?

vsoch replied:
Yes that works!! This is definitely a lot easier (and I'll update the docs). Sorry for my naivete - didn't even know I could do that :P

vsoch replied:
I'll also update the json export to use the full name of the spec with the hash, so if a user saves a json export and comes back to it later they don't need to know what `/s17m27m` means.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+
+Awesome! The above tells us that our first zlib was built without optimize (False)
+and the second was built with optimize (True). This is a small example, but there are
+actually several different types of difference properties you can view, each one being
```
Perhaps reword as `several different types of difference properties` is awkward.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+
+    $ spack diff python@2.7.8 python@3.8.8
+    ==> diff(python@2.7.8, python@3.8.8)
+    VARIANT_SET  python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8  VERSION  python Version(2.7.8)  openssl Version(1.0.2u)
```
This could grow to be quite large.  What about putting a single difference on each line?

vsoch replied:
Will do - I was using colify to follow the other commands, but this would likely be better with just newlines.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,109 @@ structured the way you want:
+        c) use `spack uninstall --all` to uninstall ALL matching specs.
+
+Oh no! This is a good use case for ``spack diff``, which can easily show us the
+"diff" or set difference between properties for two packages. Let's try it out.
```
IMO you might want to avoid mentioning `set difference` here and elsewhere since it's an "implementation detail" that may not be of interest to the average user.

vsoch committed:
real quickly that -> real quickly since

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

vsoch committed:
updating docs and print of diff

Instead of colify, we just print each diff attribute to a line.
This commit also updates the docs to show calculating a diff using
just the hashes, updates the docs to reflect this, and uses a consistent
identifier for the spec names that are diffed

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Merge branch 'develop' of github.com:spack/spack into add/spack-diff

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -728,33 +728,45 @@ real quickly that we have two!
-actually several different types of difference properties you can view, each one being
-a fact about a particular spec. The way we calculate the difference is basically by
-taking a set difference based on these facts. The first package that you provide (A)
+actually several different kinds of differences that you can view, a ``VARIANT_SET``
```
```suggestion
actually several kinds of differences that you can view, a ``VARIANT_SET``
```

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -728,33 +728,45 @@ real quickly that we have two!
+being just one of them. The first package that you provide (A)
 being diffed against B means that we see what is in A but not B. Here is another example
-with one more difference type:
+with one more difference type, ``VERSION``:
```
```suggestion
with an additional difference type, ``VERSION``:
```

tldahlgren commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+            if category != entry[0]:
+                category = entry[0]
+                rows.append(bold(category.upper()))
+            rows.append(entry[1])
```
Can you add some indentation here so the different values are offset from the bolded headings?  It'll make it a bit more readable, especially if there are a lot of differences.

```suggestion
            rows.append('  {0}'.format(entry[1]))
```

tldahlgren reviewed (changes requested):
Just a couple of small changes to mostly improve readability.

tldahlgren commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,125 @@ structured the way you want:
+    VARIANT_SET
+    python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+    VERSION
+    openssl Version(1.0.2u)
```
As I mention in a suggestion in the code for this, it would help readability if the difference output is indented at least a couple of spaces, especially when there are a lot of difference.

vsoch replied:
Agree! I'll pull the changes here and then test it locally to update the docs.

vsoch replied:
Ok, spacing is updated.

tldahlgren commented:
This is *such* a nice feature.  I've had multiple versions of a package installed when I'm testing PRs so it will be great to have a way to see what the differences between them are.

vsoch committed:
updating description of differences

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

vsoch committed:
Update lib/spack/docs/basic_usage.rst

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

vsoch committed:
adding indentation to print entries

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

vsoch committed:
updaing spacing of spack diff in docs

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@tldahlgren I totally agree! This was totally the vision of @tgamblin, I just implemented it (and will have a diff analyzer following once both are merged!)

--------------------
And I just used two "totally" in one breath. I think I lived in California too long! :laughing:

tldahlgren reviewed (dismissed):
LGTM

tldahlgren commented:
> I keep seeing this error and I'm not sure if it's transient:
> 
> ```
> 
> lib/spack/spack/test/llnl/util/tty/log.py:303: AssertionError
> ----------------------------- Captured stderr call -----------------------------
> minion: stdin.isatty(): True
> minion: ready!
> pid:        25041
> pgid:       25041
> sid:        25041
> tcgetpgrp:  25041
> 
> minion pid:  25085
> minion pgid: 25085
> minion sid:  25041
> 
> ------------------------------------------- bg
> canon: on, echo: on
> input: False
> bg: True
> 
> ------------------------------------------- fg
> canon: on, echo: on
> input: False
> bg: False
> ```
> 
> https://github.com/spack/spack/pull/22283/checks?check_run_id=2107486935

IME, this *is* transient.

tldahlgren commented:
@haampie Are you satisfied with this PR?

tgamblin commented:
I’d like to review this one too

vsoch commented:
@tgamblin one thing to discuss is when I add a diff analyzer (intended use case to save data files with diffs for the axon graveyard) how we want to organize these functions. Either we can have a diff analyzer that has all the logic imported to the command, the reverse, or something else. The analyzer is going to additionally have the weighted version metric that I made recently to generate the pretty plots.

vsoch commented:
This PR had another (what I think is) a transient error - it seems like this is the norm to have a few transient errors and have to consistently re-run.
```
.63s call     lib/spack/spack/test/cmd/extensions.py::test_extensions
=================================== FAILURES ===================================
_________________________ test_lock_debug_output[/tmp] _________________________

lock_path = '/tmp/tmp10ot3ejn/lockfile'

    def test_lock_debug_output(lock_path):
        test_debug = LockDebugOutput(lock_path)
        q1, q2 = Queue(), Queue()
>       local_multiproc_test(test_debug.p2, test_debug.p1, extra_args=(q1, q2))

lib/spack/spack/test/llnl/util/lock.py:1265: 
```

vsoch commented:
No more transient errors! We are close! :tada:

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,127 @@ structured the way you want:
+
+    $ spack diff --help
+    ...
+    --diff-type {all,version,concrete,node,node_compiler_set,node_compiler_version_set,node_os_set,node_platform_set,node_target_set,variant_set}
```
I think these names should either be marked experimental, or you, me, and maybe @becker33 and @alalazo should sit down and decide what they should be.

Right now these are taken straight out of the solver, and they don't necessarily map to anyone's intuition.  e.g., we don't really talk about `node` anywhere else in Spack -- it's just the name of the node in the spec DAG.  Maybe that one's fine. But, the `_set` suffix is an artifact of the way the solver works -- `_set` really means someone asked for that attribute to be set (but the solver can reconsider it) and the name without `_set` means "this is decided".  We probably want the latter.

It may be that we need an abstraction in here so that this (and, actually, our error messages) map to something more understandable -- and so that this can continue to exist without a lot of maintenance while the solver evolves.  One other example is that the namespace for nodes is going to be different after some upcoming changes from @alalazo.

My main concern is that people will rely on these names, so maybe for now just put a note in the output that the output of this thing is *not* going to be stable.

vsoch replied:
Maybe (along with adding the message) for this PR we should just subset to include only those that make sense? Looking [here](https://vsoch.github.io/spack-changes/data/specs/compare.html?filename=packages/ascent/ascent-spack-v0.12.0-ascent-spack-v0.14.0-comparison.json) it also looks like `node_compiler_version` includes everything from `node_compiler_set` so we can remove `node_compiler_set`. If we are comparing specs on the same platform, I'm not sure `node_platform_set` will have any value. So for this PR I propose we remove:

 - node
 - node_platform_set
 - node_compiler_set

And perhaps we could create a mapping of solver names to more human friendly names.

Another thing to note is that, per the example I linked, these lists can get very large. it's a lot to display in a terminal, so whatever strategy we choose for display, we should take this into account.

tgamblin replied:
We shouldn't remove `node` as it'll show you whether one DAG has dependencies that another doesn't -- e.g. if you had python enabled on one installation of clingo and not another, `node("python")` would show up on one side but not the other.

IMO we should use the attributes that are used to build the spec, which are listed in `display.lp`.  That's this list:

```
node/1.
depends_on/3.
version/2.
variant_value/3.
node_platform/2.
node_os/2.
node_target/2.
node_compiler/2.
node_compiler_version/3.
node_flag/3.
node_flag_compiler_default/1.
node_flag_source/2.
no_flags/2.
external_spec_selected/2.
```


The `*_set` functions are actually more like requests that happen during the solve -- they says "somebody wanted this attribute set" but they can be overridden if some higher priority thing happens -- what you want are `node_compiler` and `node_compiler_version` -- which are final decisions.

vsoch replied:
It's likely just cleaner to remove the choices group and still default to all, and people will likely start with running the command without specifying any attributes, and they can filter as needed based on what they see.

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -689,6 +689,127 @@ structured the way you want:
+
+
+This data will be much longer because along with the differences for A vs. B and
+B vs. A, we also capture the intersection.
```
this is super cool

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+import spack.cmd
+
+description = "compare two specs"
+section = "extensions"
```
This goes in `basic` (which shows up as `query packages`).  `extensions` are [something very specific](https://spack.readthedocs.io/en/latest/basic_usage.html#extensions-python-support).

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+
+description = "compare two specs"
+section = "extensions"
+level = "short"
```
I think this should go in `long`, as we try to keep only the minimal set of things in short help output (FYI this controls whether it'll show up in `spack -h` / `spack help` or in `spack -H` / `spack help --all`).

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+
+def setup_parser(subparser):
+    """Parser is only constructed so that this prints a nice help
+       message with -h. """
```
is this true?  Seems like it also provides some nice arguments like `--json` etc.?

vsoch replied:
This was the copy paste from unload and load.py, I can change it here to something else.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+            'node_compiler_version_set', 'node_os_set', 'node_platform_set',
+            'node_target_set', 'variant_set'
+        ],
+        help="select the diff type (defaults to all)"
```
See above for comments on the attributes and specifically their names.  The set of attributes displayable on a spec probably should not live here -- it currently lives in `display.lp` (to tell the solver which parts of the model to spit out) and in `spack.solver.SpecBuilder`, but only as handler methods on that class.  It should probably live in at most those two places, or at least some other place in `asp.py`

vsoch replied:
For now I'll just remove redundant attributes, and ones that aren't super meaningful (e.g., node).

vsoch replied:
Update - not removing anything.

vsoch replied:
Another update - all are totally removed (and the user can still specify as needed, which will work as they change, etc.)

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+def bold(string):
+    """Make a header string bold so we can easily see it
+    """
+    return "\033[1m" + string + "\033[0m"
```
Use Spack's color support (see `llnl.util.tty.color`), as it knows whether to disable color when you pipe output or display it on something that is not a `tty`:

```python
import llnl.util.tty.color as color
```

and, in here:

```python
return color.colorize("@*{%s}" % string)
```

`@*{}` makes things bold. You can also specify colors and things, e.g. `@c{foo}` for cyan `foo`.  We want all color to go through `llnl.util.tty.color` also so that users can control it with `spack --color={always,never, auto}`.  Also it'll be different on windows so we want that abstraction.

vsoch replied:
Ah! I almost got this when I was writing it, but didn't write the format string correctly. I'll update!

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+                rows.append(bold(category.upper()))
+            rows.append('  {0}'.format(entry[1]))
+
+    print('\n'.join(rows))
```
not a big deal but why not just print them out as you go?  I would  typically supply a stream argument for functions like this, and if it's `None`, default to `sys.stdout`.  e.g.:

```python
def print_difference(..., out=None):
    out = out or sys.stdout
...
    out.write("@*{%s}\n" % category.upper())  # print category in bold, colorized
    ...
```

Then the caller can decide how they want this printed.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+        b_name (str): the name of spec b
+        to_string (bool): return an object that can be json dumped
+    """
+    from spack.solver.asp import SpackSolverSetup
```
Can this not be imported at the top?

In general prefer `import spack.solver.asp as asp` and using `asp.SpackSolverSetup` to `from ... import`.  The former will import without needing to look at functions in the module and avoids some circular import issues.

vsoch replied:
When I was first developing it, for some reason it didn't work, but works now. I'll move it to the top!

vsoch replied:
And that totally makes sense! I was wondering why the imports are how they are.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+    }
+
+
+def to_tuple(asp_function):
```
Instead of this, can you make `AspFunction` in `asp.py` comparable as a tuple?  It should be immutable -- the `__getitem__` method on it is vestigial and can be deleted.  All you need to do is this:

```python
import llnl.util.lang as lang

@lang.key_ordering
class AspFunction(AspObject):
    def _cmp_key(self):
        return (self.name, self.args)
```

And change `self.args = []` to `self.args = ()` in the constructor (it's only ever passed tuples).

With that you should not have to call to_tuple on anything to do set comparisons, this function won't be needed, and we can do cool things with `AspFunctions` elsewhere.

vsoch replied:
cool!

vsoch replied:
Hmm, this didn't seem to work. I did remove the __getitem__ so if it breaks tests I can add it back.

tgamblin replied:
We can clean this up later.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+        help="load the first match if multiple packages match the spec"
+    )
+    subparser.add_argument(
+        '--diff-type',
```
I think this should be called `--attributes` or something like that -- as for "diff type" I usually think side-by-side, multi-way, etc.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,169 @@
+    c = compare_specs(specs[0], specs[1], to_string=True)
+
+    if args.dump_json:
+        print(sjson.dump(c))
```
probably the format for this should have something to do with the `spec.yaml` schema (this is just more on the point that we should settle on some names, at least eventually)

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1550,6 +1550,16 @@ def dag_hash_bit_prefix(self, bits):
+        cmd.display_specs but is a much simpler version to generate and return
+        a string instead of writing to an output.
+        """
+        return "%s@%s/%s" % (self.name, self.version, self.build_hash()[0:7])
```
We don't need this function -- caller should just use `spec.format("{name}{@version}{/hash:7}")`.  [See here](https://spack.readthedocs.io/en/latest/spack.html#spack.spec.Spec.format).  For this command, you can also do:

```
spec.format("{name}{@version}{/hash:7}", color=llnl.util.tty.color.get_color_when())
```

to colorize automatically (maybe `spec.format()` should use that instead of `False` for its default to further simplify -- that is likely a separate PR to fix all the places)

tgamblin reviewed (changes requested):
@vsoch: one larger question around naming that we should probably resolve, and some refactor requests and fine-tunings.

tgamblin commented:
@haampie: RE: 

> I'm thinking it would be nice to have a simple CLI menu option when spack is run in an interactive shell so the user can select a specific match. That could be used in spack uninstall when --all is not provided; as well as selecting one of the > 1 top-level specs from an environment when running spack -e my_env build-env -- bash, etc.

One of the follow-ons to this PR will be something to make those disambiguation displays show minimal diffs of the various candidates, so that you can actually see what is different *from the output*.  This PR enables that -- we'll be able to show you the variants you could type to uniquely identify the specs in the disambiguation list.

vsoch committed:
updating spack diff

We will still need to rethink how we want the solver atom names
to be presented to the user - the current set are a bit redundant
and not completely clear to the user. But we can address this in a follow
up PR, and just add a note to the user that the interface is subject
to change. This commit makes the following updates:

1. we write to an output stream as we go
2. we use the llnl color util to derive bold
3. we add a warning that the interface is subject to change
4. the diff is colorful, unless we print to json
5. diff-type argument renamed to attributes, and with added append
6. asp.SpackSolverSetup imported at top (asp module)
7. removed docstring for get_parser
8. section updated to basic from extensions

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
I suspect we need to update completion changing to basic/long

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
And let's definitely put our heads together for the console UI - the one I have here is pretty fugly.

vsoch commented:
This might be useful (never used it!) https://docs.python.org/3/library/difflib.html. here is a quick example of the html output, two diffs (each direction) side by side (you have to scroll): https://vsoch.github.io/spec-diff/

Update: the json one is really nice! https://vsoch.github.io/spec-diff/json.html

vsoch commented:
Re-opening for @alecbcs !

vsoch committed:
Merge branch 'develop' into add/spack-diff

vsoch commented:
I'm running spack style --fix locally and it's hanging on mypy checks:

```
$ bin/spack style --fix
==> Running style checks on spack
  selected: isort, mypy, flake8
==> Modified files
==> Running isort checks
  isort checks were clean
==> Running mypy checks

```

vsoch committed:
removing decorator that no longer is there?

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
ununsed function import

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
try fixing docstring

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
not sure why spack style is not being consistent

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
Does spack style actually work? I can't reproduce the issues locally and when I run `spack style --fix` I don't see additional changes. I'm not really sure what to do at this point.

vsoch commented:
@tgamblin why isn't spack style finding the same thing locally as in the PR here? Is it a version issue of isort?

tgamblin committed:
fix import order

tgamblin commented:
not sure -- is your `isort` locally bootstrapped?  if so you could try `spack clean -b` and then run it again...

tgamblin commented:
I'm using your awesome bootstrap logic from #24819

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -695,6 +695,130 @@ structured the way you want:
+and the only difference between the two is the hash. This is a good use case for 
+``spack diff``, which can easily show us the "diff" or set difference 
+between properties for two packages. Let's try it out.
+Since the only difference between these two is the hash, we provide the hashes:
```
I think this could be reworded b/c the hash isn't the only difference -- it's the variants you see below that are the only difference.

The difference in the hash tells you that there is *some* difference (b/c it's a hash of all the metadata).  You could actually see it with `spack find -v zlib`, but that's not as direct as doing `spack diff` -- and the difference might be a flag or something else so you'd have to play with `spack find` options til you find it.

Maybe the right thing to say is that the hash difference is the *only difference you see* in this `spack find` view.

vsoch replied:
Sure.

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -695,6 +695,130 @@ structured the way you want:
+
+    $ spack diff python@2.7.8 python@3.8.8
+    ==> diff(python@2.7.8/7oknfqf, python@3.8.8/vrp4fmj)
+    VARIANT_SET
```
can these be lowercase but blue to match up w/the rest of the UI (like in `spack info`

```python
import llnl.util.tty.color as color
colorized = color.colorize("@*b{%s}" % field)  # bold blue
```

vsoch replied:
Could you please make this comment on code instead of the docs for it? We do use bold and blue for info so it's not clear how to map this suggestion to where the code that does the formatting is.

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,193 @@
+        help="load the first match if multiple packages match the spec"
+    )
+    subparser.add_argument(
+        '-a',
```
this should probably be

```
'-a', '--attributes',
```

and then you can omit `dest`

tgamblin replied:
I think removing this like @vsoch suggested is the right move.

vsoch committed:
do not hard code attributes into diff command

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

alecbcs commented:
Wooo! @vsoch looks like all the tests passed!

tgamblin commented:
I've got a few more review comments but I think this is pretty close

vsoch commented:
Let's get it merged before the 5 month mark! So you have two weeks still, no rush lol

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -695,6 +695,130 @@ structured the way you want:
+      python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
+    VERSION
+      openssl Version(1.1.1j)
+      python Version(3.8.8)
```
Ok I have a slight tweak on this output that I think keeps it simple and (I think) in the spirit of this PR -- can we do something like this? (replacing `foo_set`'s with `foo`):

```diff
$ spack diff python@2.7.8 python@3.8.8
--- python@2.7.8/7oknfqf     # but show full hash here
+++ python@3.8.8/vrp4fmj     # and here
@@ variant @@
-  python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+  python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
@@ version @@
-  openssl 1.0.2u
+  openssl 1.1.1j
-  python 2.7.8
+  python 3.8.8
```

The kind of cool thing about this is it's basically `git diff` or `git -u` output and you can format it in an interpretable way (even in GitHub markdown for discussions).  We could add more formats later but I think at least people used to reading GitHub diffs would understand this.  Also we can steal console colors from `diff` like above.

Thoughts?

vsoch replied:
The reason I think we want to keep set is because that's what the facts are output as (or a subset) and that's the only way someone can then provide them as attributes. If we artificially remove set (or do any kind of edit) then a person is going to ask for a specific set of attributes (and not get them) and be really confused. And if we try to still support that, doing so means always doing a conversion between input (without set) --> input__set and then back to --> input.

tgamblin replied:
we sorted this out in https://github.com/spack/spack/pull/22283#discussion_r679545826.

Do you like the diff output though?

vsoch replied:
The json is fine (it's json!) and I liked the original output that I opened the PR with back in March, but through review here it was changed to its current form (which I don't think is great) but I'm fine with what others want. I liked the spacing and blue that I had before.

vsoch replied:
okay I added the spacing back, that makes it slightly less terrible than before :laughing: 

If you have a preference for output just tell me and I'll change it. It's just so subjective there is no way to logically choose how it looks.

tgamblin replied:
I have OCD about two things here:

1. no capitals (spack isn’t shouty and the names in the output should be what you’d provide on the command line).
2. Unified output (i.e. the individual differences are on consecutive lines like diff does it)

My main reason for wanting something like `diff` is that for large specs the format above will be hard for comparisons — you’ll have to scroll up and down to see the differences as all the diffs for the left are first, then all the fields for the right.

So… what if we do your titles (blue) but in lowercase, with actual field values unified in red/green with either < and > or - and + for left/right?

I don’t really care if the headers have @@ — it just makes the GitHub formatting look nicer bc it’s what `diff -u` does.

vsoch replied:
okay I'll have to refactor the printing so they aren't done separately then.

vsoch replied:
```diff
--- zlib@1.2.11efzjziyc3dmb5h5u5azsthgbgog5mj7g
+++ zlib@1.2.11sl7m27mzkbejtkrajigj3a3m37ygv4u2
@@ Variant_value @@
-  zlib optimize bool(False)
+  zlib optimize bool(True)
```

vsoch replied:
```diff
 bin/spack diff python@2.7.8 python@3.8.11
==> Warning: This interface is subject to change.

--- python@2.7.8/tsxdi6gl4lihp25qrm4d6nys3nypufbf
+++ python@3.8.11/yjtseru4nbpllbaxb46q7wfkyxbuvzxx
@@ version @@
-  openssl Version(1.0.2u)
+  openssl Version(1.1.1k)
-  python Version(2.7.8)
+  python Version(3.8.11)
@@ variant_value @@
-  python patches a8c52415a8b03c0e5f28b5d52ae498f7a7e602007db2b9554df28cd5685839b8
+  python patches 0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87
```

tgamblin replied:
Only nitpick is the hash should have a / before it but this looks awesome!

vsoch replied:
Oh oops, I must have deleted it when I expanded it out.

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -695,6 +695,130 @@ structured the way you want:
+    $ spack diff --help
+    ...
+      -a     {all,version,concrete,node,node_compiler_set,node_compiler_version_set,
+      node_os_set,node_platform_set,node_target_set,variant_set}
```
I like @vsoch's suggestion to remove the `-a` option and let people filter the output.

tgamblin commented:
> Let's get it merged before the 5 month mark! So you have two weeks still, no rush lol

🔥😬

vsoch committed:
tweaks to docs working for diff and attributes command

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,188 @@
+    setup = asp.SpackSolverSetup()
+
+    a_facts = set(to_tuple(t) for t in setup.spec_clauses(a))
+    b_facts = set(to_tuple(t) for t in setup.spec_clauses(b))
```
Ok I see what I was missing here.  `spec_clauses` returns `*_set` functions by default but that is likely not what we want for concrete specs.

There are two arguments to `spec_clauses` we should probably care about here:

```python
def _spec_clauses(self, spec, body=False, transitive=True):
        """Return a list of clauses for a spec mandates are true.

        Arguments:
            spec (spack.spec.Spec): the spec to analyze
            body (bool): if True, generate clauses to be used in rule bodies
                (final values) instead of rule heads (setters).
            transitive (bool): if False, don't generate clauses from
                 dependencies (default True)
        """
```

I think to keep the facts intuitive and to omit `_set`, you should pass `body=True` here.  `_set` is an implementation detail of the solver -- the "real" names that are used in concrete specs are what you get if you pass `body=True`.


I think in another PR we should add something to allow you to do `transitive=False` and only diff roots, but that is another day.

vsoch replied:
Cool I'll try that.

vsoch replied:
oh that looks much better :)

vsoch committed:
use body=True to generate facts so we do not have _set suffix

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented on lib/spack/spack/cmd/diff.py:
```diff
@@ -0,0 +1,187 @@
+        help="load the first match if multiple packages match the spec"
+    )
+    subparser.add_argument(
+        '-a', '--attributes',
```
if it's comma-separated like `-a node,variant` we should do `--attributes` -- but I think it's accumulating with append so it should be `--attribute`

vsoch committed:
rename long option --attributes -> --attribute

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
typo

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updating command line completion

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
add spacing back to spack diff output

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
refactoring spack diff to show an actual diff, and updating docs to show that

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
lowercase category names

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
update spack diff ttests for updated UI

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
add slash back for hash

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@tgamblin can we merge? :D

becker33 commented on lib/spack/spack/test/cmd/diff.py:
```diff
@@ -0,0 +1,79 @@
+    assert "variant" in diff('--first', 'mpileaks', 'mpileaks+debug')
+
+    # This matches them exactly
+    output = diff("--json", "mpileaks@2.3/ysubb76", "mpileaks@2.3/ft5qff3")
```
@tgamblin @vsoch am I reading this correctly that these hashes are "magic numbers" hardcoded in? I think we need to change this to read the hash values from the installed specs. @nhanford is running into problems where this test is not stable across a change in hash algorithm or a change the mock packages.

vsoch replied:
I don't know what you mean by magic numbers, but they refer to the same hashes you'd see on the command line with spack find and including "hash" for format:
```
$ bin/spack find --format "{name}/{hash}"
autoconf/5nymyrr2ar3userqwgewuayirjdo7c2w          ncurses/buo3tjmtuimvtofgncv6w3lzg2gmouqu
autoconf-archive/mplalc4sz2cys2wkwzji2ltyklv7x5xf  openssh/ddmvmwpu6feqyhtwnfxrsoflngvma7ll
automake/cy6bt5p74uc4qlj6jeyzuoiboaeipn5b          openssl/5m5cysfnhlgemdsngdbf4jb75cv73o7l
berkeley-db/m6cpl2epxybbnhypwvu7r3h7cq256uvg       openssl/gefi67ctqiegtpzvvnougob5oszdlysi
```
So I think if the core of how spack generates these hashes changes (@nhanford's work?) these would too and the test would also need to be updated.

becker33 replied:
Yeah, I think as part of his work Nate is going to query spack for the hashes instead of having them hardcoded, so they don't have to be maintained going forward.

tgamblin replied:
I have a fix for the test over in #25310 (still on my branch but I will push it soon). I can probably push it to Nate’s branch.

tgamblin replied:
The issue in the test isn’t the hashes here but there are some assumptions about, e.g.,  the number of attributes that differ between specs. Those change in #25310 and I bet also in @nhanford’s PR.

nhanford replied:
I fixed it in my PR and it should be pushed to my branch.

alalazo replied:
I have been trying to rebase #21683 and encountered the same issue. I think this test is bound to fail whenever the hashes change? Would it be possible to recode the test in a way to be more resilient to that?

nhanford replied:
We should have this fixed in #22845, but I need to fix a few things before we can merge that one.
